extensions/s3/src/main/java/io/deephaven/extensions/s3/S3CompletableOutputStream.java

```diff
@@ -0,0 +1,374 @@
+            throw handleS3Exception(e, String.format("initiating multipart upload for uri %s", uri), s3Instructions);
+        }
+        return response.uploadId();
+    }
+
+    /**
+     * Send a part request for the given buffer. This method assumes that the buffer is non-empty.
+     */
+    private void sendPartRequest(final OutgoingRequest request) throws IOException {
+        if (nextPartNumber > MAX_PART_NUMBER) {
+            throw new IOException("Cannot upload more than " + MAX_PART_NUMBER + " parts for uri " + uri + ", please" +
+                    " try again with a larger part size");
+        }
+        if (request.future != null) {
+            throw new IllegalStateException("Request already in progress for uri " + uri + " with part number " +
+                    nextPartNumber);
+        }
+        final UploadPartRequest uploadPartRequest = UploadPartRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .uploadId(uploadId)
+                .partNumber(nextPartNumber)
+                .build();
+        request.buffer.flip();
+        request.future = s3AsyncClient.uploadPart(uploadPartRequest,
+                AsyncRequestBody.fromByteBufferUnsafe(request.buffer));
+        request.partNumber = nextPartNumber;
+        nextPartNumber++;
+    }
+
+    private void waitForCompletion(final OutgoingRequest request) throws IOException {
+        final UploadPartResponse uploadPartResponse;
+        try {
+            uploadPartResponse = request.future.get();
+        } catch (final InterruptedException | ExecutionException e) {
+            throw handleS3Exception(e, String.format("waiting for part %d for uri %s to complete uploading",
+                    request.partNumber, uri), s3Instructions);
+        }
+        completedParts.add(SdkPojoConversionUtils.toCompletedPart(uploadPartResponse, request.partNumber));
+        request.buffer.clear();
+        request.future = null;
+        request.partNumber = INVALID_PART_NUMBER;
+    }
+
+    /**
+     * Flushes the current buffer to S3.
+     *
+     * @param force if true, forces the buffer to be flushed even if it is smaller than the minimum
+     *        {@value S3Instructions#MIN_WRITE_PART_SIZE} MiB threshold, which should only be done for the very last
+     *        part.
+     * @throws IOException if an I/O error occurs during the flush operation
+     */
+    private void flushImpl(final boolean force) throws IOException {
+        final int nextSlotId = (nextPartNumber - 1) % numConcurrentWriteParts;
+        if (pendingRequests.size() == nextSlotId) {
+            // Nothing to flush
+            return;
+        }
+        final OutgoingRequest request = pendingRequests.get(nextSlotId);
+        if (request.buffer.position() != 0
+                && request.future == null
+                && (force || request.buffer.position() >= MIN_WRITE_PART_SIZE)) {
```

Comment:
Should this be `== writePartSize`?

Reply:
AWS has the following minimum limits: https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
Basically, each part should be minimum 5 MiB, except the last part. So for flushing the data, we check if the part size is at least 5 MiB or if `force` is set, which implies the last part.

By default, we would flush if we reach the `writePartSize` but if user calls `flush`, then we just check for the minimum limit.